Added options to provider

4lexKislitsyn opened the pull request:
Added the ability to change the name of the window to search.

4lexKislitsyn committed:
Added options to configure window name to log into. Change window search mechanism.

4lexKislitsyn committed:
Removed default value and added exception.

4lexKislitsyn commented:
#4 can be fixed by using options.

poke commented on Notepad.Extensions.Logging/NotepadLogger.cs:
```diff
@@ -82,30 +84,33 @@ void WriteMessage(LogLevel logLevel, int eventId, string message, Exception ex)
-                return hwnd;
+                // when the file changes, notepad changes the name to "* Window Name", so later created loggers cannot find window
+                var builder = stringBuilderPool.Get();
+                builder.Append("*").Append(windowName);
```
This seems overly excessive for every log call. You could create the alternative window name once in the constructor and cache it as well as the `windowName` itself.

yaakov-h replied:
I also have a WIP that uses EnumWindows rather than FindWindow, which might be more efficient.

yaakov-h replied:
This has been pushed to accommodate Notepad++ support. It should be easier now to match a given window name.

poke commented on Notepad.Extensions.Logging/NotepadLoggerProvider.cs:
```diff
@@ -11,11 +11,15 @@ class NotepadLoggerProvider : ILoggerProvider
         }
 
-        public static ILoggerProvider Instance { get; } = new NotepadLoggerProvider();
+        internal NotepadLoggerProvider(NotepadProviderOptions options) : this()
```
Better use `IOptions<>` for this and make the logging options configurable through [the standard mechanism](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-3.1#configure-logging). You can use `LoggerProviderOptions.RegisterProviderOptions` and the `ProviderAliasAttribute` for that.

4lexKislitsyn replied:
FIxed. Now options can be configured through [configure method](https://docs.microsoft.com/ru-ru/dotnet/api/microsoft.extensions.dependencyinjection.optionsconfigurationservicecollectionextensions.configure?view=dotnet-plat-ext-3.1).

poke commented on Notepad.Extensions.Logging/NotepadProviderOptions.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <summary>
+        /// Default options for <see cref="NotepadLoggerProvider"/>.
+        /// </summary>
+        public static NotepadProviderOptions Default { get; } = new NotepadProviderOptions
+        {
+            WindowName = "Untitled - Notepad",
+        };
```
This likely doesn’t do what you expect it to do. This will essentially just help establishing a singleton instance for the options which is then shared globally. And you *mutate* that instance when you call `optionsConfiguration?.Invoke(options);` in the extension method.

A better way would be to just assign this default directly to `WindowName` and create a new instance in the `AddNotepad` extension method. Or use the options pattern and don’t worry about instantiation.

4lexKislitsyn replied:
Yes, it was my mistake, I just hurried up to create PR 😄

4lexKislitsyn replied:
It's outdated. Now supports default behaviour without window name.

yaakov-h commented on Notepad.Extensions.Logging/NotepadLogger.cs:
```diff
@@ -82,30 +84,33 @@ void WriteMessage(LogLevel logLevel, int eventId, string message, Exception ex)
+                // when the file changes, notepad changes the name to "* Window Name", so later created loggers cannot find window
+                var builder = stringBuilderPool.Get();
+                builder.Append("*").Append(windowName);
+                hwnd = NativeMethods.FindWindow(null, builder.ToString());
```
using stringbuilder is overkill for a single character concatenation

4lexKislitsyn committed:
Removed string builder from logger. Added support of IOptions.

4lexKislitsyn committed:
Merge from source repository.

4lexKislitsyn commented:
Can we make `WindowFinder` non static to pass window name?

And in Notepad++ you can rename tab without saving it. But regex doesn't allow us to use this function. It will be nice to implement such scenario:
1. Create new tab 
1. Rename it to "logging"
1. in config set window name "logging"

But compile regex every time isn't good way, maybe regex can be changed to something like `@"- Notepad\+\+$"`?

yaakov-h commented:
> And in Notepad++ you can rename tab without saving it.

Ha, I did not know that.

With regards to customization, my current thinking is:

- If the user specifies a window name, then we find a window matching that name (with or without the `*` prefix). If they don't specify a window name, we should probably have the value as null, and then use the current behaviour as a default.
- We need a way to figure out if that windows is Notepad or NP++. I guess they could specify it, but autodetection would be nicer - possibly from the remote process's main module name?

yaakov-h commented on Notepad.Extensions.Logging/NotepadLogger.cs:
```diff
@@ -12,14 +12,18 @@ namespace Notepad.Extensions.Logging
             this.stringBuilderPool = stringBuilderPool;
             this.categoryName = categoryName;
+            this.windowName = windowName ?? throw new ArgumentNullException(nameof(windowName));
+            this.changedWindowName = $"*{windowName}";
```
I'd rather just skip the `*` when searching (e.g. https://github.com/yaakov-h/Notepad.Extensions.Logging/blob/master/Notepad.Extensions.Logging/WindowFinder.cs#L45-L49) than keep track of it separately.

4lexKislitsyn replied:
Yes, since you added another search method it can be removed.

yaakov-h commented on Notepad.Extensions.Logging/NotepadLoggerProvider.cs:
```diff
@@ -1,25 +1,46 @@
+            optionsReloadToken?.Dispose();
+        }
+
+        private void ReloadLoggerOptions(NotepadLoggerOptions options)
```
Omit `private` since its the default.

4lexKislitsyn replied:
Fixed. I'm used to define this because of the code style of the current work.

yaakov-h commented on Notepad.Extensions.Logging/NotepadLoggerProvider.cs:
```diff
@@ -1,25 +1,46 @@
 
 namespace Notepad.Extensions.Logging
 {
+    [ProviderAlias("Notepad")]
```
I'm starting to wonder if `Notepad` and `Notepad++` should be two separate loggers (though with shared code, of course)

4lexKislitsyn replied:
Different `ILogger` classes or `ILoggerProvider`? I think it should be single provider, that can be configured to different applications kind. So it will writes only in one application.

poke replied:
I think it should be very unlikely that someone would want to log to _both_ because that's what two loggers would imply. I think it would be better to just add support for it in one logger.

However regarding Notepad++, since this is a more common tool for developers to use (compared to notepad), maybe logging to it automatically this would be a bit troublesome and distracting. So maybe we should look for a more explicit opt-in mechanism. For example requiring users to name the window in a specific way.

4lexKislitsyn replied:
Some naming format will be realy helpfull for Notepad++. It will allow us to change regex on something like `logging(?:-(\S+))? - Notepad\+\+` and if user provide name to search it should be equal to the first group value.

yaakov-h replied:
From a developer experience perspective I think it would be easier for users to call:

```csharp
AddNotepad(o => o.DocumentName = "MyDocumentName");
```

or:

```csharp
AddNotepadPlusPlus(o => o.DocumentName = "MyDocumentName")
```

and then for us to automatically match it to the window title (e.g. `*MyDocumentName - Notepad`).

And if they don't specify a document name, then we look for the defaults - `Untitled` for Notepad, and `/new \d+/` for NP++.

For now its probably simpler to have one logger that supports both (i.e. just `AddNotepad`), though.

4lexKislitsyn replied:
It can be solved by adding internal property to `NotepadLoggerOptions` and calling configure method.
```csharp
public static ILoggingBuilder AddNotepad(this ILoggingBuilder builder)
    => AddProvider(builder, WindowKind.Notepad);

public static ILoggingBuilder AddNotepad(this ILoggingBuilder builder, Action<NotepadLoggerOptions> configure)
    => AddProvider(builder, WindowKind.Notepad, configure);

public static ILoggingBuilder AddNotepadPlusPlus(this ILoggingBuilder builder)
    => AddProvider(builder, WindowKind.NotepadPlusPlus);

public static ILoggingBuilder AddNotepadPlusPlus(this ILoggingBuilder builder, Action<NotepadLoggerOptions> configure)
    => AddProvider(builder, WindowKind.NotepadPlusPlus, configure);

public static ILoggingBuilder AddProvider(this ILoggingBuilder builder, WindowKind kind, Action<NotepadLoggerOptions> configure)
{
    if (configure == null)
    {
        throw new ArgumentNullException(nameof(configure));
    }
    AddProvider(builder, kind);
    builder.Services.Configure(configure);
    return builder;
}

static ILoggingBuilder AddProvider(ILoggingBuilder builder, WindowKind kind)
{
    builder.AddConfiguration();

    builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider, NotepadLoggerProvider>());
    LoggerProviderOptions.RegisterProviderOptions<NotepadLoggerOptions, NotepadLoggerProvider>(builder.Services);

    if (kind != default)
    {
        builder.Services.Configure<NotepadLoggerOptions>(opt => opt.Application = kind);
    }

    return builder;
}
```

And later use this property for search like window name. What do you think? Of course it can be done after this PR as next enhancment.

4lexKislitsyn replied:
> From a developer experience perspective I think it would be easier for users to call

In this way we will unable to support multilanguage. In some cultures windows uses dash character instead of minus (`"–".Equals("-")` is false). So provide only document name will break multilanguage support for notepad. Notepad++ will work well.

yaakov-h replied:
Good thinking. I'm probably gonna have to build up a list then of how Notepad[++] builds its title bar in every language... 😞

yaakov-h commented:
`WindowFinder` should probably remain static, unless/until I can figure out how to pass a managed .NET object by references as the state to `EnumWindows`. That would clean up the threadstatics.

4lexKislitsyn commented:
> unless/until I can figure out how to pass a managed .NET object

Till it I can add static string to `WindowFinder` and change it in `ReloadLoggerOptions`. Is it ok?

yaakov-h commented:
I've pushed some changes to remove the statics, which should make it easier to pass the necessary options around.

4lexKislitsyn committed:
Merge non static changes.

4lexKislitsyn committed:
Added support of window name from options in WindowEnumerationState class.

4lexKislitsyn commented:
I've pushed updates. I changed regex to define window kind. What do you think about some rules for Notepad++ window's name? We talked about it [here](https://github.com/yaakov-h/Notepad.Extensions.Logging/pull/5#discussion_r429102685).

4lexKislitsyn committed:
Removed private keyword.

yaakov-h commented on Notepad.Extensions.Logging/WindowEnumerationState.cs:
```diff
@@ -47,26 +49,50 @@ public bool ExamineWindow(IntPtr hWnd)
         }
 
-        static Regex notepadPlusPlusRegex = new Regex(@"^new \d+ - Notepad\+\+$", RegexOptions.Compiled);
+        static Regex notepadPlusPlusRegex = new Regex(@" - Notepad\+\+$", RegexOptions.Compiled);
```
This regex can be replaced with a simple `.EndsWith` call.

yaakov-h commented on Notepad.Extensions.Logging/WindowEnumerationState.cs:
```diff
@@ -47,26 +49,50 @@ public bool ExamineWindow(IntPtr hWnd)
-                WindowKind = WindowKind.NotepadPlusPlus;
-                Handle = NativeMethods.FindWindowEx(Handle, IntPtr.Zero, "Scintilla", null);
-                return true;
+                switch (titleText)
```
Having a `switch` seems odd when there's only one case.. `if/else if/ else if` would probably be better here.

4lexKislitsyn replied:
Ok, I will change it.

yaakov-h commented on Notepad.Extensions.Logging/WindowFinder.cs:
```diff
@@ -12,11 +12,14 @@ public WindowFinder()
             var stateObject = statePool.Get();
             try
             {
+                stateObject.WindowName = WindowName;
```
Can we make this a method parameter of `FindNotepadWindow` instead of a public property of `WindowFinder`?

4lexKislitsyn replied:
If will force us to pass it into `NotepadLogger`. I think if logger get instance of `IWindowFinder` it should not be able to change window name.

If you think that logger should manage window finder behaviour I will change it.

yaakov-h replied:
I don't see a problem with `NotepadLogger` reading the window name from the options and passing it to the finder. Ultimately the logger needs to find the window and write to it...

4lexKislitsyn committed:
Moved window name to IWindowFinder method parameter from property.

4lexKislitsyn commented:
I've pushed updates:
- made window name parameter of method instead of property
- replaced `switch` with `if/else if/else if`
- returned regex
- removed default value for window name

Let me know if I should make any changes to apply the PR.

I created another [branch](https://github.com/4lexKislitsyn/Notepad.Extensions.Logging/tree/dev/app-kind) where I added simple support of document name when window name is null or whitespace. If you want I can add this branch to the PR.

yaakov-h commented on Notepad.Extensions.Logging/WindowEnumerationState.cs:
```diff
@@ -16,11 +16,13 @@ public WindowEnumerationState()
 
         public IntPtr Handle { get; private set; }
         public WindowKind WindowKind { get; private set; }
+        public string WindowName { get; internal set; }
```
not `private set`?

yaakov-h commented on Notepad.Extensions.Logging/WindowFinder.cs:
```diff
@@ -12,11 +12,12 @@ public WindowFinder()
             var stateObject = statePool.Get();
             try
             {
+                stateObject.WindowName = windowName;
```
oh I see, we need to set it here.
